### DISCUSSION TITLE
Unexhaustive pattern match on dataclass field union

### DISCUSSION THREAD

**Discussion by ducdetronquito** [microsoft/pyright#9429] [Q&A] [answered]
Hello !

Here is a minimal case of pattern matching on a dataclass field that is the union of two types:

```python
from dataclasses import dataclass

@dataclass
class Classic:
    pass


@dataclass
class Admin:
    pass


@dataclass
class User:
    name: str
    role: Classic | Admin


some_user = User("pluto", role=Admin())

match some_user:
    case User(_, role=Classic()):
        pass
    case User(_, role=Admin()):
        pass
```

I would expect that the pattern match is exhaustive here but pyright (1.1.386) disagree:

```
error: cases within match statement do not exhaustively handle all values
    Unhandled type: "User"
``` 

Pyright only considers that the pattern match is exhaustive when I add the following branch:

```python
match some_user:
    case User(_, role=Classic()):
        pass
    case User(_, role=Admin()):
        pass
    case User():
        pass
```

Do you have any idea why this 3rd branch is required ?

Thanks in advance, and have a nice day :palm_tree: 

EDIT: [Here is a link to the same code on pyright's playground](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4QGcmwMZIFABmUA9gLYgAmAhkldgDZXroCm6IMpYxUSlNdRszx4AAtVoMm6PFOYgAwkPQxsKPCE3hpIsRME657AIIVSMAHbqt24SPECjs5SACqrKNa0WqpFmkwoDS0Sen9FZVUQAB8QU3MLXXQyFgB9AFcPEABeNw8ACgAiMHp0pGJCgBoQUJZs%2BMt8gEomkVIabAQMFIyPL01sJhY8lih81Ora7KVpVWam-ptbGRtB1hGxiZriMPqzRpbFmzAdIA)

**Comment by erictraut** [comment, marked as answer]
Exhaustion detection uses type narrowing to determine whether you have handled all possible types. In this case, the subject type starts as `User`. The subsequent case statements do not narrow the type further. The type of the subject remains `User` after the first and the second. There is no notion in the Python type system of the type `User where the "role" attribute refers to a "Classic" object`. 

If you want static analysis exhaustion to work, you'll need to change your code somewhat.

Code sample in [pyright playground](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4QGcmwMZIFB4BmUA9gLYgAmAhktdgDbXroCm6IMZYJUSVtekxYEAAjTqNm6PFJYgAwsPQxsKPCE3hpYiUJ1yOAQUpkYAO3VbtIvOMGHZykAFU2UANoAVNEumqQAB8QEzNzAF0rLXNqMlY0TCgNLVIGeJAvAkpWQhBCAFdzbAAKdHJWAH189zQ3Vk8-FlVwoNd3D1CLcIBKKM0yWmwEDHKqmuTrTWxmVjb64oqAGhBU1gBeRpUS7t6JyeswHX2pmbmoBeXVtc7zYp2%2B481DkSA)

```python
# pyright: strict

from dataclasses import dataclass

@dataclass
class Classic:
    pass

@dataclass
class Admin:
    pass

@dataclass
class User[T: Classic | Admin]:
    name: str
    role: T

def func(some_user: User[Classic] | User[Admin]):
    match some_user:
        case User(_, role=Classic()):
            pass
        case User(_, role=Admin()):
            pass

```

**Reply by ducdetronquito** [reply]
Thanks a lot for your detailed answer :)
